app/services/onlinetesting/Phase3TestService.scala

```diff
@@ -217,20 +217,31 @@ trait Phase3TestService extends OnlineTestService with Phase3TestConcern {
       }
     }
 
-  def markAsCompleted(launchpadInviteId: String)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = eventSink {
-    phase3TestRepo.getTestGroupByToken(launchpadInviteId).flatMap { test =>
-      if (test.testGroup.tests.find(_.token == launchpadInviteId).get.completedDateTime.isEmpty) {
-        for {
-          _ <- phase3TestRepo.updateTestCompletionTime(launchpadInviteId, dateTimeFactory.nowLocalTimeZone)
-          updated <- phase3TestRepo.getTestGroupByToken(launchpadInviteId)
-          _ <- phase3TestRepo.updateProgressStatus(updated.applicationId, ProgressStatuses.PHASE3_TESTS_COMPLETED)
-        } yield {
-          AuditEvents.VideoInterviewCompleted(updated.applicationId) ::
-            DataStoreEvents.VideoInterviewCompleted(updated.applicationId) ::
-            Nil
+  def markAsCompleted(launchpadInviteId: String)(implicit hc: HeaderCarrier, rh: RequestHeader): Future[Unit] = {
+
+    // Launchpad only: If a user has completed unexpire them
+    def removeExpiryIfSet(applicationId: String, expired: Boolean) = expired match {
```

Comment:
I'd say the control coupling was a bad thing if the method wasn't an inner method of this class, it really only exists to keep the for comprehension tidy - so being control coupled was almost exactly it's purpose - i'm not sure i can see a way to rewrite this without putting an if() inside the for comp, or switching out to a big nested flatmap - what would you suggest?